File: kotlin/04-supervision/src/main/kotlin/org/elu/kotlin/akka/Monitoring.kt
package org.elu.kotlin.akka

import akka.actor.AbstractActor
import akka.actor.ActorRef
import akka.actor.ActorSystem
import akka.actor.Props
import akka.actor.Terminated
import akka.japi.pf.ReceiveBuilder

class Ares(private val athena: ActorRef) : AbstractActor() {
  override fun preStart() {
    context.watch(athena)
  }

  override fun postStop() {
    println("Ares postStop...")
  }

  override fun createReceive(): Receive =
      ReceiveBuilder()
          .match(Any::class.java) { when(it) { is Terminated -> context.stop(self) } }
          .build()
}

class Athena : AbstractActor() {
  override fun createReceive(): Receive =
      ReceiveBuilder()
          .match(String::class.java) {
            println("Athena received $it")
            context.stop(self)
          }
          .build()
}

fun main() {
  val system = ActorSystem.create("supervision")

  val athena = system.actorOf(Props.create(Athena::class.java), "athena")
  system.actorOf(Props.create(Ares::class.java, athena), "ares")

  println("starting...")
  Thread.sleep(100)
  athena.tell("hello", ActorRef.noSender())

  Thread.sleep(1000)
  println("stopping...")

  system.terminate()
}

File: scala/04-supervision/src/main/scala/org/elu/scala/akka/Monitoring.scala
package org.elu.scala.akka

import akka.actor.{Actor, ActorRef, ActorSystem, Props, Terminated}

class Ares(athena: ActorRef) extends Actor {
  override def preStart(): Unit = {
    context.watch(athena)
  }

  override def postStop(): Unit = {
    println("Ares postStop...")
  }

  override def receive: Receive = {
    case Terminated =>
      context.stop(self)
  }
}

class Athena extends Actor {
  override def receive: Receive = {
    case msg =>
      println(s"Athena received $msg")
      context.stop(self)
  }
}

object Monitoring extends App {
  val system = ActorSystem("monitoring")

  val athena = system.actorOf(Props[Athena], "athena")
  val ares = system.actorOf(Props(new Ares(athena)), "ares")

  println("starting...")
  Thread.sleep(100)
  athena ! "hello"

  Thread.sleep(100)
  println("stopping...")

  system.terminate()
}
